BootloaderCorePkg/Library/S3SaveRestoreLib/S3SaveRestore.c

```diff
@@ -0,0 +1,180 @@
+      Ptr += CommHdr->TotalSize;
+      continue;
+    }
+    break;
+  }
+
+  CommHdr = (BL_PLD_COMM_HDR *) DataPtr;
+  if ( Ptr + CommHdr->TotalSize > SmmBase + SmmSize ) {
+    DEBUG ((DEBUG_ERROR, "Overflowing SMM region\n"));
+    return;
+  }
+  CopyMem ((VOID *)Ptr, DataPtr, CommHdr->TotalSize);
+}
+
+
+/**
+  Get pointer to a specific struct area in TSEG
+
+  @param      Id                Id of the struct to be retreived
+
+  @retval     NULL              Struct with signature not found
+  @retval     Ptr               Pointer to struct with signature
+**/
+VOID *
+EFIAPI
+GetS3Info (
+  IN  UINT8   Id
+  )
+{
+  UINT8             *Ptr;
+  LDR_SMM_INFO      LdrSmmInfo;
+  UINT8             *SmmBase;
+  UINT32            SmmSize;
+  BOOLEAN           FoundInfo = FALSE;
+  BL_PLD_COMM_HDR   *CommHdr;
+
+  PlatformUpdateHobInfo (&gSmmInformationGuid, &LdrSmmInfo);
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  if (LdrSmmInfo.Flags & SMM_FLAGS_4KB_COMMUNICATION) {
+    SmmSize = SIZE_4KB;
+  } else {
+    DEBUG((DEBUG_ERROR, "Invalid Communication area size\n"));
+    return NULL;
+  }
+
+  for (Ptr = SmmBase; Ptr < SmmBase + SmmSize; ) {
+    CommHdr = (BL_PLD_COMM_HDR *) Ptr;
+    if ( CommHdr->Signature == BL_PLD_COMM_SIG) {
+      if ( CommHdr->Id != Id) {
+        Ptr += CommHdr->TotalSize;
+        continue;
+      }
+      FoundInfo = TRUE;
+    }
+    break;
+  }
+
+  if (!FoundInfo) {
+    return NULL;
+  }
+
+  return (VOID *) Ptr;
+}
+
+
+/**
+  This function restores the states of the registers that were
+  set to be saved by the bootloader in the normal boot path.
+  After the payload has updated the 'Val' field of the reg info
+  in the normal boot path, we simply OR the updated 'Val'
+  with the existing vale of the register in the S3 resume boot path.
+  This function is only called in the S3 resume path.
+
+  @param    S3SaveReg   S3_SAVE_REG info offset
+    
+  @retval   None
+
+**/
+VOID
+EFIAPI
+RestoreS3RegInfo (
+  IN  S3_SAVE_REG   *S3SaveReg
+  )
+{
+  UINT8                 Index;
+  UINT32                Data32;
+  UINT16                Data16;
+
+  if (S3SaveReg == NULL || S3SaveReg->S3SaveHdr.Id != S3_SAVE_REG_COMM_ID) {
+    DEBUG ((DEBUG_ERROR, "Invalid S3 Save Reg pointer\n"));
+    return;
+  }
+
+  for (Index = 0; Index < S3SaveReg->S3SaveHdr.Count; Index++) {
+    if (S3SaveReg->RegInfo[Index].Addr != 0x00) {
+      //
+      // Check register type and width
+      //
+      if ( ((S3SaveReg->RegInfo[Index].Type != IO) && (S3SaveReg->RegInfo[Index].Type != MEM)) ||
+           ((S3SaveReg->RegInfo[Index].Width != WIDE32) && (S3SaveReg->RegInfo[Index].Width != WIDE16)) ) {
+        DEBUG ((DEBUG_ERROR, "ERROR: not supported regtype (0x%x) or regwidth (0x%x)\n",\
+        S3SaveReg->RegInfo[Index].Type, S3SaveReg->RegInfo[Index].Width));
+        return ;
+      }
+      //
+      // Restore registers saved by payload
+      //
+      if (S3SaveReg->RegInfo[Index].Width == WIDE16) {
```

Comment:
Better to use macro.

ex)
#define RESTORE_S3_REGS(TYPE, SIZE, ADDR, VALUE) \
  TYPE##Or##SIZE(ADDR, (UINT##SIZE)VALUE); \
  (VOID)TYPE##Read##SIZE(ADDR)

RESTORE_S3_REGS (type == Mem ? Mmio : Io, Width * 8, Addr, Val);